### PR TITLE
storage: apply snapshot in its own batch

### PR DESCRIPTION
**Pull request by tbg**
As discussed in https://reviewable.io/reviews/cockroachdb/cockroach/6961#-KJ7BJC50LpDnmQTOW7j

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7135)

### CODE REVIEW

**Comment by tamird** [comment]
:lgtm:

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> #### storage: apply snapshot in its own batch
> 
> As discussed in https://reviewable.io/reviews/cockroachdb/cockroach/6961#-KJ7BJC50LpDnmQTOW7j
> 
> 
> 
</details>


---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135#-:-KJpkt1twTjoFmlBKJBh:1417088721)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135#-KJpkmPJK-fNj1NJo45T:-KJpkmPJK-fNj1NJo45U:-710549126) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3757642e36c62f6a506014da9dcf54211519fa9/storage/replica.go#L1407)):*
> ```Go
> 	}
> 	rd := r.mu.raftGroup.Ready()
> 	lastIndex := r.mu.lastIndex
> ```

Is this being used now?

---

*[storage/replica.go, line 1425 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135#-KJpkxCPe3v50beuUU_i:-KJpkxCQJkwa3NCEb-Ly:1963977497) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3757642e36c62f6a506014da9dcf54211519fa9/storage/replica.go#L1425)):*
> ```Go
> 		}
> 		r.mu.Lock()
> 		lastIndex = r.mu.lastIndex
> ```

Is this needed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135#-KJpkmPJK-fNj1NJo45T:-KJpleRBjDOIqycDvkFd:1987922230) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3757642e36c62f6a506014da9dcf54211519fa9/storage/replica.go#L1407)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Is this being used now?
</details>
in the non-snapshot case, looks like it's used in the append call

---

*[storage/replica.go, line 1425 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135#-KJpkxCPe3v50beuUU_i:-KJplcLmBptsY8RdmSFm:1750797834) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3757642e36c62f6a506014da9dcf54211519fa9/storage/replica.go#L1425)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Is this needed?
</details>
it's used in the append call below

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135)*

**Comment by tschottdorf** [comment]
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135#-KJpkmPJK-fNj1NJo45T:-KJploJ9DI8yK41LCtnm:370391589) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3757642e36c62f6a506014da9dcf54211519fa9/storage/replica.go#L1407)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> in the non-snapshot case, looks like it's used in the append call
</details>
Yep. Added comments since it's a bit hard to parse.

---

*[storage/replica.go, line 1425 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135#-KJpkxCPe3v50beuUU_i:-KJplrNp-yNSmEtIftO1:-1933320192) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3757642e36c62f6a506014da9dcf54211519fa9/storage/replica.go#L1425)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> it's used in the append call below
</details>
Yep. Comment added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7135)*
